src/parser/Quads.cpp

```diff
@@ -0,0 +1,96 @@
+  return std::visit([](const auto& graph) -> T { return graph; }, graph);
+};
+
+// ____________________________________________________________________________________
+std::vector<SparqlTripleSimpleWithGraph> Quads::toTriplesWithGraph() const {
+  std::vector<SparqlTripleSimpleWithGraph> quads;
+  size_t numTriplesInGraphs = std::accumulate(
+      graphTriples_.begin(), graphTriples_.end(), 0,
+      [](size_t acc, const GraphBlock& block) {
+        return acc + get<ad_utility::sparql_types::Triples>(block).size();
+      });
+  quads.reserve(numTriplesInGraphs + freeTriples_.size());
+  ad_utility::appendVector(
+      quads, transformTriplesTemplate(freeTriples_, std::monostate{}));
+  for (const auto& [graph, triples] : graphTriples_) {
+    ad_utility::appendVector(
+        quads,
+        transformTriplesTemplate(
+            triples, expandVariant<SparqlTripleSimpleWithGraph::Graph>(graph)));
+  }
+  return quads;
+}
+
+// ____________________________________________________________________________________
+std::vector<parsedQuery::GraphPatternOperation>
+Quads::toGraphPatternOperations() const {
+  auto toSparqlTriple = [](const std::array<GraphTerm, 3>& triple) {
+    return SparqlTriple::fromSimple(SparqlTripleSimple(
+        triple[0].toTripleComponent(), triple[1].toTripleComponent(),
+        triple[2].toTripleComponent()));
+  };
+
+  using namespace parsedQuery;
+  std::vector<GraphPatternOperation> operations;
+  operations.emplace_back(
+      BasicGraphPattern{ad_utility::transform(freeTriples_, toSparqlTriple)});
+
+  for (const auto& [graph, triples] : graphTriples_) {
+    // We need a `GroupGraphPattern` where the graph is set. This contains the
+    // triples inside another `GraphPattern`.
+    GraphPattern tripleSubPattern;
+    tripleSubPattern._graphPatterns.emplace_back(
+        BasicGraphPattern{ad_utility::transform(triples, toSparqlTriple)});
+    operations.emplace_back(
+        GroupGraphPattern{std::move(tripleSubPattern),
+                          expandVariant<GroupGraphPattern::GraphSpec>(graph)});
+  }
+  return operations;
+}
+
+// ____________________________________________________________________________________
+void Quads::forAllVariables(std::function<void(const Variable&)> f) {
+  auto visitGraphTerm = [&f](const GraphTerm& t) {
+    if (std::holds_alternative<Variable>(t)) {
+      f(std::get<Variable>(t));
+    }
+  };
+  auto visitTriple = [&visitGraphTerm](const std::array<GraphTerm, 3>& triple) {
+    const auto& [s, p, o] = triple;
+    visitGraphTerm(s);
+    visitGraphTerm(p);
+    visitGraphTerm(o);
+  };
+  auto visitGraphBlock = [&visitTriple, &f](const GraphBlock& block) {
+    const auto& [graph, triples] = block;
+    if (holds_alternative<Variable>(graph)) {
+      f(get<Variable>(graph));
+    }
+    ql::ranges::for_each(triples, visitTriple);
+  };
+  ql::ranges::for_each(graphTriples_, visitGraphBlock);
+  ql::ranges::for_each(freeTriples_, visitTriple);
+}
```

Comment:
This explicit function now can also have an explicit unit test that handles all the cases.